ENH: Update scalar representations as per NEP 51

seberg opened the pull request:
This is a WIP for implementing NEP 51/gh-22261.  There are a few things that go beyond just changing the scalar representation:

1. It makes `get_formatter()` a more central/public API, to also move `fmt=...` to it.
   * This introduces the distinction that if the `dtype` is known, we can e.g. represent a longdouble as `'3.1'` (with quotes) rather than `np.longdouble('3.1')`.
   * `fmt` is mainly used for `fmt="r"` (previous point), I think `fmt="s"` should be supported, but always be identical to `str(arr[0])`.
   * There could be an argument for the `fmt` path to be a ufunc (it should not inspect array values, unlike the default printing).
   * `fmt` should be used more cleanly for the `__format__` work.  This should be fine, I think it only means pushing the "option" adjustments lower.

2. `arr.tofile()` used `repr(arr.item(...))` to print each element.  This is weird/problematic:
   * For longdouble, I want that repr to be `np.longdouble('3.1')`.
   * For strings, this means it uses `'string'` or `b'string'`

3. Record and MA printing has to be adjusted to use the new functionality.  This affects record scalars and the printing of the masked array `fill_value`.

   After some back and forth, I think the default should just be to go back to `str()` (this was changed silently in gh-6199).  `str(arr.item())` is not ideal for `float32` for example.
   This could use some thought (or should change to use the above `get_formatter(fmt="s")`, but I hope this can be decoupled from NEP 51.

---

I plan to update the NEP 51 draft to include these points.  Happy to discuss here of course (will make pushing NEP 51 quicker).  But for now the draft-PR exists primarily to help with the NEP 51 discussion.

**TODOS::**

* [x] Refguide check is failing (and may lead to fixups being necessary?)
* [x] Discuss and accept NEP 51
* [x] Decide whether this should be hidden behind a flag or if it should interact with the `legacy` print-mode (i.e. to allow inclusion in the next release, but only switch with NEP 50 – which would likely be a NumPy 2.0 as well)
* [ ] New tests for some paths are needed.  Mainly `arr.tofile()` is not well tested (which is why the change here should not fail tests).

ganesh-k13 committed:
ENH: Changed repr of np.bool_

ganesh-k13 committed:
ENH: Used raw boolean value

ganesh-k13 committed:
ENH: Refactored to use PyUnicode_FromString

ganesh-k13 committed:
ENH: Fixed doc-string

ganesh-k13 committed:
ENH: Added release notes (#17592)

ganesh-k13 committed:
ENH: Fixed release notes `np.bool_` repr (#17592)

eric-wieser committed:
Update doc/release/upcoming_changes/17592.improvement.rst

ganesh-k13 committed:
ENH: Changed repr of ints

ganesh-k13 committed:
ENH: Changed repr of ints to use tp_name

seberg committed:
WIP: Get to the NEP 51 state (and try to figure that out...)

seberg committed:
TST: Fixup tests for repr changes

seberg committed:
TST: Some more test fixups for repr changes

still quite a few to do, and there are some things that we need to
figure out...

seberg committed:
TST: STrengthen test on newer Python versions

It seems older Python versions rounded `str` sometimes, I do not
think this is strictly true anymore.
IF this is wrong on certain systems, we need to rethink some things!

seberg committed:
Fixup tests, repr, and guard against using full repr in array printing

seberg committed:
WIP: Continue with refactoring more...

seberg committed:
WIP: Refactor array formatter with a new `get_formatter`

seberg committed:
ENH: Fixup records and void scalar printing

seberg committed:
ENH: Only print type information when helpful for MA fill value

seberg committed:
API: Switch `ndarray.tofile` back to defaulting to `str()` usage

`tofile()` should probably be improved in either case, but this changes
it back to using `str()` in order to fix storing of longdouble numbers.
Note that, it still goes via Python float for float32, which is OK
but maybe not ideal (stores more digit than useful, which may be misleading).

This also changes `object` arrays, and fixes(?) string/bytes arrays since
these would previously print with quotes or `b''` prefix (which seems just
undesired).

The use of `repr` may have been useful/desired on Python 2 for floats, but
Python 3 makes no real distinction between both.

seberg committed:
TST: Fixup tests to make windows/32bit systems and linter happy

tylerjereddy commented:
@seberg so far so good for latest `main` of SciPy alongside this feature branch in my hands (Ubuntu Linux; i9-7900X):

- `python dev.py test -j 10`
- `37669 passed, 2222 skipped, 139 xfailed, 9 xpassed in 83.59s (0:01:23)`

seberg committed:
Adapt format to not block s and r format codes

Not sure that is really necessary, but I think empty string for
`s` should be OK in practice (in principle `!s` exists in formatting
syntax which calls `str` explicitly...)

seberg commented:
So, I realized that the empty format string `""` already has more or less the meaning of `str()`, and in that sense I was thinking of using the `repr` function itself as a singleton for "scalar repr" (rather than `"r"`).

I don't really have an opinion, could even say that `str` should be used for `!s` formatting effectively.  Or could go back, because why would anyone use `{obj:r}` for something that is not a repr...

seberg committed:
WIP: Fixup arrayprint for scalar values

seberg committed:
MAINT: Switch to also have str explicitly, fixup MA and forward str/repr to scalar

Most importantly, use the scalar `str()` for formatting str/repr of
"array elements" when a format code is given (for floats).
The reason is that floats to print differently currently and falling back is
simplest.

seberg committed:
DOC: Update reference to pass refguide

This makes things pass even though there should be a massive amount
of types added in other places, so that it is not complete.

But I think this may be useful, since the complete change would be
quite verbose...

One thing to note may be that `np.float64(nan)` currently does not
round-trip correctly of course, this would require adding `np.`
to the repr (or quotes).

seberg committed:
STY: Fix linter

seberg committed:
BUG: Fixups for last changes to make CI pass

I had tested the refguide mainly (the functions part only), which missed
a lot of smaller errors in the last iteration...

seberg committed:
DOC: First small refdoc update (very verbose update required, too large for PR)

seberg committed:
DOC: Remove outdated release note (eventually need new referencing NEP)

seberg committed:
DOC: Fixup docs; only "big" changes not all scalars

As said before, all scalars would be noisy, while this gives at
least an idea of what is off (and makes CI happy)

h-vetinari commented:
I saw https://github.com/numpy/numpy/pull/22261 only now, so I won't complain if this is too late, but I'd really like to get away from `long`, `longlong`, etc. I'd much prefer to just use sized integers throughout. It's a much saner model, and now that C99 is available on all platforms, we could do it (modulo the fact that LAPACK and perhaps some other venerable libs still have those types in their API 😑)

seberg commented:
I am proposing to print the bit-sized version, even if imprecise sometimes.  Yes, it would be nice to change it, but it seems like a big ABI change on the C-side?

h-vetinari commented on doc/source/reference/arrays.classes.rst:
```diff
@@ -665,9 +665,9 @@ objects as inputs and returns an iterator that returns tuples
 ...     print(val)
-(1, 0)
-(0, 1)
-(2, 0)
-(3, 1)
+(np.int64(1), np.int64(0))
+(np.int64(0), np.int64(1))
+(np.int64(2), np.int64(0))
+(np.int64(3), np.int64(1))
```
Didn't the text of NEP 51 say:
> The NEP does not propose to change how these scalars print – only
their representation (``__repr__``) will be changed.

seberg replied:
Tuples print the `repr` of their content.

There already is `.item()` to convert to the python scalar where applicable.  The only problem is longdouble, which should be rare.

About changing how this is backed: I am happy for someone to think about it, but it is neither this, nor will I seriously consider breaking ABI.  Propose it as an API break and I will be interested.

h-vetinari commented:
> I am proposing to print the bit-sized version, even if imprecise sometimes.

Looked more closely, looks fine.

> Yes, it would be nice to change it, but it seems like a big ABI change on the C-side?

I'm afraid so, but I still think it'd be worth it _eventually_  (numpy 2.0 material?).

h-vetinari reviewed (commented):
There seems to be a mismatch between the implementation and the NEP w.r.t. whether printing is affected or not.

Based on the implementation, I think this PR has pretty high chance of disrupting services where something gets calculated with numpy, and then passed on in some serialized form to another consumer, where numpy (or even python) might not necessarily be installed.

Is there any fallback we could provide so that people can adapt such interfaces (e.g. `.value` or `.raw` or ...)?

ev-br commented:
> the refguide check does not seem to notice changes from 0.0 to np.float64(0.0)

It won't, as long as `np.allclose(eval('np.float64(0.0)'), 0.0)`

charris commented:
Needs rebase.

seberg committed:
fix rebase error

seberg committed:
MAINT: Allow all options again (because subarrays need them) and fix test

If we use a format string, just ignore all other options and grab
the global ones.  I am not 100% sure if there might be a weird path
where things get lost on the way for structured arrays or so,
this change is new, I so I think its fine eithe  way.

(and I suspect it is actually correct)

seberg committed:
DOC: More doc fixes to pass tests

This does *not* fix everything, I am hoping for an automated tester
based on doctestplus for this (I actually have that mostly ready)

Doing this here woul dbe quite spammy though!

pllim commented:
@mhvk , how do you think `astropy` can/should deal with this in our doctest downstream? I can only think of disabling doctest for numpy-dev and then when numpy 2.0 is released, we update the docs and disable doctest for numpy 1.x. However, this comes with the risk that we won't catch any incompatibility if a feature is only tested in the doc examples. Hope you can advise. Thanks! 🙏

mhvk commented:
Hmm, a quick test shows
```
======================= 455 failed, 23250 passed, 3669 skipped, 240 xfailed in 375.49s (0:06:15) =======================
```
with essentially none of the documentation passing.

I think the tests are most easily adjusted; often, it seems a matter of using `str` or `!s` instead of `repr` and `!r`, and it needs adjusting for the new `get_formatter`.

For the docs, I think we may need to update `doctest-plus` to automatically ignore `np.float64` etc. @seberg - this would be something that would be nice for numpy to provide...

mhvk commented:
For `MaskedArray`, perhaps we should use `str` for `fill_value`. E.g., currently,
```
masked_array(data=['2000-01-01 00:00:00.000', --],
             mask=[False,  True],
       fill_value=np.str_('N/A'),
            dtype='<U23')
```
I think if the `dtype` of `fill_value` is the same as the array dtype, it should not be shown.

seberg commented:
> I think if the dtype of fill_value is the same as the array dtype, it should not be shown.

That is already the case, I thought.  But strings are awkward because the default fill value has a smaller string length.

I had worked on automatically _fixing_ the docs, I guess auto ignoring any scalar constructor shouldn't be super hard, but is it vital to have to push forward here?

mhvk commented:
@seberg - if there is a way to automatically fix the docs, that would be awesome! Seeing the number of failures, I think that would save a lot of people a lot of pain.

That said, for most of our failures, it is things like `q=1*u.m; q.value`, which used to give `1.0` and now gives `np.float64(1.0)` -- I'm not sure we shouldn't just move to returning the python float or no longer decaying the underlying array to a scalar.

mhvk commented:
p.s. It is really unfortunate that doctests, which in principle are a nice idea, have made changes in representation so painful to make!

mhvk commented on numpy/ma/core.py:
```diff
@@ -4067,7 +4067,18 @@ def __repr__(self):
+
+        if self._fill_value is None:
+            self.fill_value  # initialize fill_value
+        if self._fill_value.dtype == self.dtype:
```
To avoid `np.str_('N/A')` maybe
```
if (self._fill_value.dtype == self.dtype
        or self.dtype.kind in 'US' and self.dtype.kind == self._fill_value.dtype.kind):
```

seberg replied:
sure why not, done.  Reall, it's nonesense probably that they even _can_ diverge, but since strings are usually longr than the typical `'N/A'` fill value, I don't mind guessing this.

mhvk commented:
p.p.s. Just re-read NEP 51 - clearly, everything is there (and it remains a good idea!). I would suggest that we special-case string `fill_value` in `np.ma.MaskedArray`; see in-line comment...

seberg committed:
MAINT: Special case string fill-value in MA repr

Just to remove unnecessary visual clutter in this special case
where it is almost always no additional information.

seberg commented:
There are decisions to be made about how you want to adopt it into the docstrings.  I suspect we should bite the bullet and just use `np.float64(1.0)`, etc. but if you don't, then you need a doctest plugin to remove the scalar repr maybe.

The question remains a bit if it really matters so much to restrict doctests to <2.0 until the docs are fixed and then (hopfully soon) restrict it to >=2.0?  (sooner would be nice as the new versions are really more relevant).

Unlike normal tests, you don't want to add version guards to the docs reasonably, so changes in behavior in principle always come with such problems?

seberg committed:
DOC: Add a release note for NEP 51 related changes

mhvk commented:
Having the legacy option seems a good idea. Also, who knows, maybe there is going to be a 1.26...

seberg committed:
ENH; Add legacy fallback mode for non NEP 51 printing because...

seberg committed:
Allow legacy fallback (fix float16, heh)

seberg commented:
OK, pushed that, few tests (contributiosn would be helpful) could be nice (also a small bug fix).  But the astropy test suite is being helped by it is probably the better test anyway.

1.26 must happen for 3.12 support in some capacity, but we already removed things from the C-API with significant improvements relying on it, so you already would have to cherry-pick what you like (or very carefully undo other commits).

I am honestly -1 to any serious 1.26 release without a narrow focus.  There is exactlye *one* thine I want in 2.0, and this the weak promotion, because I think its widely wanted.  Everything else, I would prefer to do a 3.0 release soon, rather than fussing about delaying a 2.0.

Sure, NEP 50 is hard enough, I am willing to even drop that.  My point is, delaying will solve nothing unless the horizon is *very* clear.  We should have done 2.0 5 years ago and we could be at 4.0 by now.

seberg committed:
TST,MAINT: Add tests for both new and old scalar repr and fix complex

seberg committed:
DOC: Mention `np.set_printoptions` in the release note

seberg committed:
TST: Ensure little-endian int est

seberg commented:
I ran the astropy test suite, it seems `repr` is also used somewhere in YAML saving?  With the legacy options, the only problem it faces is that `_get_format_function` is renamed/different.
Of course the tests passing is a lie, because the YAML related issues are probably real (from wherever they actually originate); setting the print option for non-doctests isn't really right.

mhvk commented:
I already have a PR in the works for astropy to deal with this. IIRC it mostly meant changing things from `!r` to `!s`. I'll recheck...

mhvk commented:
See https://github.com/astropy/astropy/pull/15065 - as I remembered, the actual code changes are minimal, and I guess for the docs we'll just have to bite the bullet...

mhvk committed:
Simplify, removing new fmt option

mhvk committed:
Minimize changes relative to main

mhvk reviewed (commented):
@seberg - I had another look and made a lot of comments, but then was struck by thinking that the new `fmt` option really is weird - `arrayprint` already has a way to override the formatting via `formatter` so why provide another?

I think it would be better to make this a minimally invasive PR and not touch `arrayprint` more than is necessary: really, what is needed here is to allow quoting of long doubles, and to ensure structured voids print reasonably.

I made a PR to your PR that removes the complications, and that still does seem to pass the tests (at least the basic ones; I really need to try to understand how to do testing in the new setup; are there instructions anywhere?). All the complexity is now in `_void_scalar_repr`.

Note that I also don't understand why `where` is being changed...

seberg commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -3320,7 +3320,7 @@ array_set_datetimeparse_function(PyObject *NPY_UNUSED(self),
 PyArray_Where(PyObject *condition, PyObject *x, PyObject *y)
 {
-    PyArrayObject *arr, *ax, *ay;
+    PyArrayObject *arr, *ax, *ay = NULL;
```
ouch, sorry must have mistakenly committed :(.

seberg committed:
Remove unnecessary legacy check and longdouble quoting support

Undoing the other stuff, basically means that quoting was now used
where it is unclear that it should be (in array printing/repr) but
not where it definitely should be (in void repr).

So there is just no point, we basically decided to not support it
for now and consider it a bug to be fixed later.

seberg committed:
MAINT: Undo bad rebase (or maybe small accidental commit)

seberg committed:
Address some other review comments

seberg commented:
After talking with Marten, I simply undid (in practice) *all* array printing changes.

The basic thing is that we need a concepts for "item repr" and "item scalar", probably including "exact item repr" and "exact item scalar".
1. Add quoting for longdoubles in `np.void((longdouble, longdouble), dtype="...")`.
   * My code did add quotes on longdoubles in the default array repr.  I had the machienry to change it, but it wasn't quite correct:  Since the default array repr/printing truncates precision, there is no point in adding quotes (it is not an exact repr!).
3. Add quoting around datetimes in `repr(array)` but ideally *not* in `str(array)` (dates are clear enough).

After removing the machinery to do that, there was no point keeping manual longdouble quoting support which would be harder to use in the place where you want it (void repr) and was used in the place where we definitely don't want it (`str(array)`).

seberg committed:
Hack strings to be correct for fill-value

seberg committed:
Update release note and add note to NEP that it was not fully implemented

seberg committed:
MAINT: Tweak string check (forgot the kind, but maybe tuple is nice)

seberg committed:
Also force repr for object

mhvk commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -1045,3 +1049,48 @@ def test_ctx_mgr_as_smth(self):
+    assert repr(scalar) == representation
+
+    np.set_printoptions(legacy="1.25")
+    assert repr(scalar) == legacy_repr
```
Again would use context manager.

mhvk commented on numpy/core/tests/test_longdouble.py:
```diff
@@ -40,13 +40,13 @@ def test_scalar_extraction():
 def test_repr_roundtrip():
```
Change name to `test_str_roundtrip()`

mhvk commented on numpy/core/tests/test_longdouble.py:
```diff
@@ -40,13 +40,13 @@ def test_scalar_extraction():
 
 
 @pytest.mark.skipif(string_to_longdouble_inaccurate, reason="Need strtold_l")
 def test_repr_roundtrip_bytes():
```
Same.

mhvk commented on numpy/core/tests/test_longdouble.py:
```diff
@@ -277,7 +277,7 @@ def test_tofile_roundtrip(self):
 
 def test_repr_exact():
```
Again, should now be `test_str_exact`

mhvk commented on numpy/core/tests/test_longdouble.py:
```diff
@@ -316,7 +316,7 @@ class TestCommaDecimalPointLocale(CommaDecimalPointLocale):
 
     def test_repr_roundtrip_foreign(self):
```
And here too.

mhvk commented on numpy/core/tests/test_records.py:
```diff
@@ -132,7 +132,9 @@ def test_0d_recarray_repr(self):
+            "np.record((1, 2.0, '2003'), "
+            "dtype=[('f0', '<i4'), ('f1', '<f8'), ('f2', '<M8[Y]')])")
         # 1.13 converted to python scalars before the repr
         try:
```
Another place where it could be `with np.printoptions(...)`

seberg replied:
There is a lot of this, not going to touch it here.  Also, I would hope we can get rid of 1.13?

mhvk replied:
OK, makes sense.

For 2.0, yes, seems to make sense to get rid of 1.13.

mhvk replied:
(though not in this PR!)

mhvk commented on numpy/ma/tests/test_core.py:
```diff
@@ -556,23 +556,23 @@ def test_str_repr(self):
               mask=[[False, False, False],
                     [False,  True, False]],
-              fill_value=999999,
+              fill_value={np.array(999999)[()]!r},
```
Why not just put `np.float64(999999.0)` here (and below)? Really would hope we're not changing it again!

seberg replied:
Because this is an integer and this is needed because it's either int64 or int32

mhvk replied:
Ah, darn.

mhvk commented on numpy/ma/core.py:
```diff
@@ -4067,7 +4067,22 @@ def __repr__(self):
+                and self.dtype.kind == self._fill_value.dtype.kind):
+            # Allow strings: "N/A" has length 3 so would mismatch.
+            fill_repr = repr(self.fill_value.item())
+        elif self._fill_value.dtype == self.dtype and not self.dtype == object:
```
I think indeed this is OK for now!

mhvk commented on doc/neps/nep-0051-scalar-representation.rst:
```diff
@@ -232,6 +232,14 @@ found `here <https://github.com/numpy/numpy/pull/22449>`_
 ==============
 
+.. note::
+    This part has *not* been implemented with the initial changes.
```
Might as well give a PR number, for easier reference: gh-22449.

mhvk reviewed (commented):
@seberg - I think this all makes sense now. The only slightly larger remaining comment is about `repr_format` and `str_format` -- see there.

It may be good to have a follow-up issue about remaining items. It seems to be current `array2string` is actually fine, but the question seems to be whether there should be some kind of option where all scalars, including `void`, would round-trip. If we want to support sub-arrays for `void`, then I think this would imply that having a round-trip version for `array` would be a simple extension. But it seems to me it would also be fine *not* to do that - the main goal, of ensuring scalars are recognizable, has been reached here. Anyway, for a separate issue, let's get this one in first!

seberg commented on numpy/core/arrayprint.py:
```diff
@@ -395,9 +399,13 @@ def _object_format(o):
     return fmt.format(o)
 
 def repr_format(x):
+    if isinstance(x, (np.str_, np.bytes_)):
```
On the one hand, yeah. On the other hand, this is the drop-through for user defined dtypes and I don't know if it is wise to call `.item()` on those.

mhvk replied:
Ah, good catch, I had not realized that `numpystr` is used as a default for `dtype.type is None`.

Since really the point of my earlier comments was to keep the changes minimal, it makes sense to stick with your choice - though perhaps add a TODO that really this is to be fixed? 

Also, would it work to make the change in `repr_format`? Is `str_format` used? Indeed, for `str_format`, is there actually any difference between `str(x)` and `str(x.item())`?

seberg committed:
MAINT: A few small fixups from review

seberg commented:
To be honest, I still think we will eventually go back to what I had, although hopefully it's OK to dump the special special fmt states (and improved) (and just have `r` and `s`, which void can use, whatever `subarray` makes of that, I don't care if it checks the printoptions and doesn't round-trip).

Structured void repr should stay integrated with array repr, I think.  So any changes there need to be integrated into `array2string` internals (which are a mess anyway, and in need of reasonable hooks for both user dtypes and format string passing).

seberg commented:
Can we push this over the finish line?  Yes, my doctest changeset isn't quite there yet, but not sure it will be helpful to wait, I will expect that the worst affected projects would use the `legacy=` option anyway initially?

mhvk reviewed (approved):
@seberg - sorry, lots of other stuff came up. But yes, this is good to go as far as I'm concerned!

charris commented:
Thanks Sebastian.

larsoner commented:
I can see a lot of projects hitting this... I'm trying this workaround in `mne/conftest.py` in `pytest_configure` after reading the release notes from here:
```
    try:
        np.set_printoptions(legacy="1.25")
    except Exception:
        pass  # probably missing kwarg
```
I'll let people know in 1h if it works!

larsoner commented:
So far I had to change it to `legacy=125`, using a string broke stuff *later* (not at `set_printoptions` time) for released versions of NumPy. I'll update in another hour...

larsoner commented:
Looks like we're headed to green with `legacy="1.25"` (set only when the NumPy version is > 1.25), so the suggested workaround is all good (when deployed correctly) :+1:

seberg commented:
I will note that just setting the legacy mode for the tests will unfortunately hide issues if you use `repr(scalar)` for serialization/writing to a file (or a dependency does).  So it is a good hot-fix for doctests or tests which check `repr` explicitly, but might cause real issues to be missed.
